src/ekat/util/ekat_meta_utils.hpp

```diff
@@ -0,0 +1,235 @@
+#ifndef EKAT_META_UTILS_HPP
+#define EKAT_META_UTILS_HPP
+
+#include <tuple>
+#include <type_traits>
+
+namespace ekat
+{
+
+// A parameter pack. To be used for template specializations
+template<typename... Ts>
+struct TypeList;
+
+template<typename T>
+struct TypeList<T> {
+  using head = T;
+  static constexpr int size = 1;
+};
+
+template<typename T, typename... Ts>
+struct TypeList<T,Ts...> {
+private:
+  using tail = TypeList<Ts...>;
+public:
+  using head = T;
+  static constexpr int size = 1 + tail::size;
+};
+
+// Cat two packs
+template<typename P1, typename P2>
+struct CatLists;
+
+template<typename... T1s, typename... T2s>
+struct CatLists<TypeList<T1s...>,TypeList<T2s...>> {
+  using type = TypeList<T1s...,T2s...>;
+};
+
+// Find position of first occurrence of type T in a TypeList, store in 'pos'.
+// If not found, set pos = -1.
+template<typename T, typename TypeList>
+struct FirstOf;
+
+template<typename T, typename H>
+struct FirstOf<T,TypeList<H>> {
+  static constexpr int pos = std::is_same<H,T>::value ? 0 : -1;
+};
+
+template<typename T, typename H, typename... Ts>
+struct FirstOf<T,TypeList<H,Ts...>> {
+private:
+  static constexpr int tail_pos = FirstOf<T,TypeList<Ts...>>::pos;
+public:
+  static constexpr int pos = std::is_same<H,T>::value ? 0 : (tail_pos>=0 ? 1+tail_pos : -1);
+};
+
+// Check if a TypeList contains unique types
```

Comment:
If this is supposed to be a meta-predicate, what do you think about sticking the word `Is` in there somewhere? Like `IsUniqueTypeLIst`?